account_edi_ubl_cii_purchase_match/models/account_edi_xml_ubl_20.py

```diff
@@ -0,0 +1,111 @@
+            return res
+        invoice_origin_node = tree.find("./{*}OrderReference/{*}ID")
+        if invoice_origin_node is None:
+            return res
+        order_ref = invoice_origin_node.text
+        self._match_invoice_to_purchase_order(invoice, order_ref)
+        return res
+
+    def _match_invoice_to_purchase_order(self, invoice, order_ref):
+        purchase_order = self.env["purchase.order"].search(
+            [
+                "|",
+                ("name", "=", order_ref),
+                ("partner_ref", "=", order_ref),
+                ("state", "in", ("purchase", "done")),
+            ],
+            limit=1,
+        )
+        if not purchase_order:
+            return False
+        for invoice_line in invoice.invoice_line_ids:
+            self._match_invoice_line_to_purchase_order_line(
+                invoice_line, purchase_order.order_line
+            )
+        return True
+
+    def _match_invoice_line_to_purchase_order_line(self, invoice_line, purchase_lines):
+        if invoice_line.purchase_line_id:
+            return False
+        product = self._get_matching_product(invoice_line)
+        purchase_line = purchase_lines.filtered(lambda line: line.product_id == product)
+        invoice_line._set_product(product)
+        invoice_line.purchase_line_id = purchase_line
+        return True
+
+    def _get_matching_product(self, invoice_line):
+        product_model = self.env["product.product"]
+        if invoice_line.product_id:
+            return invoice_line.product_id
+        product_name = invoice_line.name
+        if not product_name:
+            return product_model
+        product = product_model.search([("name", "=", product_name)])
+        if product:
+            return product
+        partner = invoice_line.move_id.partner_id
+        if not partner:
+            return product_model
+        supplierinfo = self.env["product.supplierinfo"].search(
+            [
+                ("product_name", "=", product_name),
+                ("partner_id", "=", invoice_line.move_id.partner_id.id),
+            ],
+            limit=1,
+        )
+        if not supplierinfo:
+            return product_model
+        if supplierinfo.product_id:
+            return supplierinfo.product_id
+        if (
+            supplierinfo.product_tmpl_id
+            and len(supplierinfo.product_tmpl_id.product_variant_ids) == 1
+        ):
+            return supplierinfo.product_tmpl_id.product_variant_ids
+        return product_model
+
+    def _import_fill_invoice_line_form(
+        self, journal, tree, invoice, invoice_line, qty_factor
+    ):
+        res = super()._import_fill_invoice_line_form(
+            journal, tree, invoice, invoice_line, qty_factor
+        )
+        supplier_product_code = self._find_value(
+            "./cac:Item/cac:SellersItemIdentification/cbc:ID", tree
+        )
+        if invoice_line.product_id or not supplier_product_code:
```

Comment:
There is I think a bug in the call to super that retrieves the product. It's searching for a product with default_code based on the SellersItemIdentification while it should be BuyersItemIdentification
https://github.com/odoo/odoo/blob/f07949e4ccdb5d7a17a6678b008fa6785ec3fb12/addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py#L727
I would force a new call to the method `_retrieve_product` with the right argument